lib/elixir/lib/calendar/date.ex

```diff
@@ -757,6 +769,75 @@ defmodule Date do
     end
   end
 
+  @doc """
+  Shifts a date by given units. Raises an `ArgumentError` if the
+  given date is not a valid Calendar.ISO date.
+
+  Available shift units are: `:year, :month, :week, :day` and the shift
+  is always applied in that order.
+
+  `Date.shift/2` defines a week as 7 days and a year as 12 months.
+
+  When shifting by month:
+  - it will shift to the current day of a month
+  - if the current day does not exist in a month, it will shift to the last day of a month
+
+  ## Examples
+
+      iex> Date.shift(~D[2016-01-03], month: 2)
+      ~D[2016-03-03]
+      iex> Date.shift(~D[2016-02-29], month: 1)
+      ~D[2016-03-29]
+      iex> Date.shift(~D[2016-01-31], month: 1)
+      ~D[2016-02-29]
+      iex> Date.shift(~D[2016-01-31], year: 4, day: 1)
+      ~D[2020-02-01]
+
+  """
+  # @doc since: "1.5.0"
+  @spec shift(Calendar.date(), [shift_unit()]) :: t
```

Comment:
Something else to consider is that, we will eventually want to add this function to `DateTime` and in there we can shift to an ambiguous date. We need to decide if we will either resolve the conflict automatically or return an error. If we return an error, then it means the consistent return type for `shift` across all modules would be a `{:ok, _} or {:error, _}` tuple.

Reply:
Fair point, i mentioned that in my PR description as well. Once you're open to proceeding with this approach in general, i'd be leaning towards not auto-resolving ambiguous `DateTime`s. As soon as an intermediate shift fails, it should return an error tuple, unless it's the last one, in which case we'd just return the amb_gap three-element tuple.

`DateTime.shift/3` could receive an option to auto-resolve, considering either the datetime before or after though.

Let me know what you think? Either way it makes sense to me to consistently return tuples in Date and NaiveDateTime as well, even though we won't deal with ambiguous datetimes there.

Reply:
> I was looking into using Calendar.months_in_year/1 which works well for the month shift as long as the date is not shifted by more months than a year has (unless it's the ISO calendar, then it always works afaik :)). 

If we make this support multiple calendars, the whole logic would be inside the calendar. This function would pretty much just validate the input and call `Calendar.shift_date/4` (and `Calendar.shift_naive_date_time/7` for the others).